content/blog/2022-01-15-moving-local-experiments-to-the-cloud-with-iterative-terraform-provider.md

```diff
@@ -0,0 +1,280 @@
+  However, try to make it unique, it shouldn’t be in conflict with names of
+  existing cloud resources.
+- _cloud_ (**required**): cloud provider to run the task on. This can be aws,
+  gcp, az, or k8s.
+- _region_: you can choose the region where the compute resources should be
+  allocated.
+- _machine_: if you know the exact kind of machine that you'd like to use, you
+  can specify it here. Alternatively, the Iterative Provider offers some common
+  machine types which are roughly the same for all supported clouds. For
+  example,
+  [l+k80](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#l+k80)
+  stands for a "Large, with (at least) 12 CPU cores, 112 GB of RAM and 2 GPU
+  devices".
+- _directory_: Specify a directory on your local machine relative to your
+  project folder which you would like to sync with the remote machine. This way
+  you can share your whole project or parts of it with a remote machine. In my
+  example, I am sharing the whole project.
+- _script_ (**required)**: This is where the Iterative Provider's magic happens,
+  i.e. this is where we define a script that should be run on a provisioned
+  machine.
+
+Tke a look at the script. In the simplest scenario, all we need to do on a new
+machine to run the training script is to set up the Python environment with
+required libraries, and run the script. And that's exactly what the script here
+does (granted you have a `requirements.txt` file). If you simply want to train
+your model on a machine with more memory, this may be enough. However, if you
+want your training code to leverage GPUs, the script will look a bit more
+complex.
+
+## Training with GPU
+
+Setting up the environment to leverage GPU devices is a bit involved as you need
+to make sure you have the required drivers, NVIDIA-docker, and CUDA.
+
+```
+resource "iterative_task" "tpi-examples-gpu" {
+    name      = "tpi-examples-gpu"
+    cloud     = "aws"
+    region    = "us-east-2"
+    machine   = "l+k80"
+    directory = "."
+
+    script = <<-END
+    #!/bin/bash
+    sudo apt update
+    sudo apt-get install -y software-properties-common build-essential ubuntu-drivers-common
+    sudo ubuntu-drivers autoinstall
+
+    sudo curl -fsSL https://get.docker.com -o get-docker.sh && sudo sh get-docker.sh &&
+    sudo usermod -aG docker ubuntu
+    sudo setfacl --modify user:ubuntu:rw /var/run/docker.sock
+    curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
+    curl -s -L https://nvidia.github.io/nvidia-docker/ubuntu20.04/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
+    sudo apt update && sudo apt install -y nvidia-docker2
+    sudo systemctl restart docker
+    rm get-docker.sh
+
+    nvidia-smi
+    docker run --rm --gpus all -v "$PWD:/tpi" iterativeai/cml:0-dvc2-base1-gpu \
+        /bin/bash -c "cd /tpi; pip install -r requirements.txt; python src/train.py"
```

Comment:
Unfortunately in your suggestion, GPU isn't used, the training happens using CPU because some libraries are not found (probably related to CUDA). I can try to experiment and make it work without Docker, but it might get rather complex.

Reply:
It's worth it to remove the Docker complexity from this blog. Nicer to leave Docker for a follow-up post.

In terms of how to install CUDA here just do it the same way we do for the CML docker image?

We base it on [cuda11.0.3-ubuntu20.04-base](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/11.0.3/ubuntu2004/base/Dockerfile) + [runtime](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/11.0.3/ubuntu2004/runtime/Dockerfile) + [cudnn8](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/11.0.3/ubuntu2004/runtime/cudnn8/Dockerfile)... defines all the `apt` packages needed. You'll also likely need:

```bash
ENV PATH /usr/local/nvidia/bin:/usr/local/cuda/bin:${PATH}
ENV LD_LIBRARY_PATH /usr/local/nvidia/lib:/usr/local/nvidia/lib64
```

Reply:
Addressed in b0443d1e629d77d7b44cf9b3c9f444f80629913a by switching to using an AMI